### PR TITLE
[SPARK-36793][K8S] Support write container stdout/stderr to file

### PR DESCRIPTION
**Pull request by warrenzhu25**
### What changes were proposed in this pull request?
Support write container stdout/stderr to file

### Why are the changes needed?
If users want to sidecar logging agent to send stdout/stderr to external log storage,  only way is to change entrypoint.sh, which might break compatibility with community version.

### Does this PR introduce _any_ user-facing change?
Yes. User can enable this feature by spark config.

### How was this patch tested?
Added UT in BasicDriverFeatureStepSuite and BasicExecutorFeatureStepSuite

### CODE REVIEW

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh [hunk 1]
```diff
@@ -107,4 +107,14 @@ case "$1" in
 esac
 
 # Execute the container CMD under tini for better hygiene
-exec /usr/bin/tini -s -- "${CMD[@]}"
+if [ -z "$K8S_SPARK_LOG_PATH" ]; then
```
Comment (martin-g) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh:110]: The env var has been renamed to `SPARK_K8S_LOG_PATH`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala [hunk 2]
```diff
@@ -82,7 +82,14 @@ private[spark] class BasicDriverFeatureStep(conf: KubernetesDriverConf)
           .withName(env._1)
           .withValue(env._2)
           .build()
-      }
+      } ++ {
+      if (conf.get(KUBERNETES_LOG_TO_FILE)) {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala:86]: Indentation?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala [hunk 3]
```diff
@@ -177,7 +177,14 @@ private[spark] class BasicExecutorFeatureStep(
             .withValue(opt)
             .build()
         }
-      }
+      } ++ {
+      if (kubernetesConf.get(KUBERNETES_LOG_TO_FILE)) {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:181]: ditto.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala [hunk 4]
```diff
@@ -441,6 +441,19 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
     ))
   }
 
+  test("container log configs") {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala:444]: Could you add a JIRA ID prefix like 
```
- test("container log configs") {
+ test("SPARK-36793: container log configs") {
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you for making a PR, @warrenzhu25 .
However, without a rolling feature, this will kill the executor due to OutOfDisk.

**Comment by warrenzhu25** [comment]
> Thank you for making a PR, @warrenzhu25 . However, without a rolling feature, this will kill the executor due to OutOfDisk.

@dongjoon-hyun thanks for reviewing this. For log rolling feature, could we support this in the future? I can create a jira ticket to track this. This is similar issue as shuffle output, which also might run out of disk.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Please note that there are many Spark streaming jobs which run forever. Without a proper fix, this proposal is an invalid option in the production environment.

> @dongjoon-hyun thanks for reviewing this. For log rolling feature, could we support this in the future? I can create a jira ticket to track this. This is similar issue as shuffle output, which also might run out of disk.

Unfortunately, I'll cast -1 for this AS-IS proposal.

**Comment by dongjoon-hyun** [comment]
In addition, the following is totally different. Users are able to control the shuffle data lifecycle programmatically while this PR doesn't allow any option.
> This is similar issue as shuffle output, which also might run out of disk.

**Comment by martin-g** [comment]
One can use external tools like [logrotate](https://linux.die.net/man/8/logrotate) for this.

**Comment by dongjoon-hyun** [comment]
@warrenzhu25 I'm going to close this PR because this level of implementation is too primitive and can be achieved via the custom `entrypoint.sh` file when the user builds their docker image. Please feel free to reopen this when there is an update on the proposal.

**Pull request closed without merging**